File: java-streams/example/src/main/java/streams/MapFlatMap.java
package streams;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class MapFlatMap {

    List<Employee> findEmployees(List<Person> persons) {

        return persons.stream()
                .map(person -> new Employee(person.getFirstName() + " " + person.getLastName()))
                .collect(Collectors.toList());
    }

    List<Address> findAddresses(List<Person> persons) {

        return persons.stream()
                .flatMap(person -> person.getAddresses().stream())
                .collect(Collectors.toList());
    }

    List<String> nameParts(List<Person> persons) {

        return persons.stream()
                .flatMap(person -> Stream.of(person.getFirstName(), person.getLastName()))
                .collect(Collectors.toList());
    }
}

File: java-streams/examples/kotlin/src/main/kotlin/streams/MapFlatMap.kt
package streams

class MapFlatMap {

    internal fun findEmployees(persons: List<Person>) =
            persons.map { person -> Employee(person.firstName + " " + person.lastName) }

    internal fun findAddresses(persons: List<Person>) =
            persons.flatMap { person -> person.addresses }

    internal fun nameParts(persons: List<Person>) =
            persons.flatMap { (_, firstName, lastName) -> listOf(firstName, lastName) }
}
